src/DGtal/base/FunctorHolder.h

```diff
@@ -0,0 +1,291 @@
+    }
+
+  /** @brief Writes/Displays the object on an output stream.
+   *
+   * @param out the output stream where the object is written.
+   */
+  inline
+  void selfDisplay ( std::ostream & out ) const
+    {
+      out << "[FunctorHolder]";
+      if ( SharedPtrTrait::apply(myFunctor) )
+        out << " using std::shared_ptr storage (rvalue)";
+      else if ( ReferenceWrapperTrait::apply(myFunctor) )
+        out << " using std::reference_wrapper storage (lvalue)";
+      else
+        out << " using custom storage";
+    }
+
+  /** @brief Checks the validity/consistency of the object.
+   *
+   * @return 'true' if the object is valid, 'false' otherwise.
+   */
+  inline constexpr
+  bool isValid() const
+    {
+      return true;
+    }
+}; // End of class FunctorHolder
+
+/** @brief Overloads 'operator<<' for displaying objects of class @ref FunctorHolder.
+ *
+ * @param out the output stream where the object is written.
+ * @param object the object of class @ref FunctorHolder to write.
+ * @return the output stream after the writing.
+ */
+template <typename FunctorStorage, bool NeedDereference>
+std::ostream&
+operator<< ( std::ostream & out, const FunctorHolder<FunctorStorage, NeedDereference> & object )
+{
+  object.selfDisplay( out );
+  return out;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Implementation details
+namespace {
+
+  /// Overload that stores the object in a std::reference_wrapper when given object is a lvalue reference.
+  template <typename Function>
+  inline auto holdFunctorImpl(Function && fn, std::true_type)
+      -> FunctorHolder<decltype(std::ref(std::forward<Function>(fn))), false>
+    {
+      return FunctorHolder<decltype(std::ref(std::forward<Function>(fn))), false>{ std::forward<Function>(fn) };
+    }
+
+  /// Overload that stores the object in a std::shared_ptr when given object isn't a lvalue reference.
+  template <typename Function>
+  inline auto holdFunctorImpl(Function && fn, std::false_type)
+      -> FunctorHolder<std::shared_ptr<Function>, true>
+  {
+      return FunctorHolder<std::shared_ptr<Function>, true>{ std::make_shared<Function>(std::forward<Function>(fn)) };
+  }
+
+} // anonymous namespace
+
+
+//////////////////////////////////////////////////////////////////////////////
+// FunctorHolder construction helper
+/** @brief Hold any callable object (function, functor, lambda, ...) as a C(Unary)Functor model.
+ *
+ * @tparam  Function  The callable object type (auto-deduced, do not specify it)
+ * @param   fn        The callable object to be holded.
+ * @return  a FunctorHolder instance with appropriate storage type depending on the given object.
+ *
+ * @see FunctorHolder and @ref moduleFunctors for more informations
+ */
+template <typename Function>
+inline auto holdFunctor(Function && fn)
+  -> decltype(holdFunctorImpl(std::forward<Function>(fn), typename std::is_lvalue_reference<Function>{})) // Can we please pass to C++14 ?
```

Comment:
So with C++14 you just need to specify auto and it deduces the return type from the return statement ?

Reply:
yep, see comment above :wink: